### PR TITLE
Tweak Map/Hashtable interfaces

### PR DESCRIPTION
**Pull request by rgrinberg**
As discussed on slack:

* The old `add` is now `set
* There's now a safe `add` called `add_exn`. It raises if we try to overwrite a key.

We should create an `add` as well, but I'm not sure of the type signature. Perhaps:

```ocaml
module Duplicate : sig
  type 'a t
  val make : 'a -> 'a t
  val get : 'a t -> 'a
end
val add : t -> key -> 'a -> (t, 'a Duplicate.t) Result.t
```

We can't use base because it uses polymorphic variants. And I'd like something that works with result. This makes it much more usable with the new syntax.

### CODE REVIEW

File: src/stdune/hashtbl.ml [hunk 1]
```diff
@@ -58,6 +58,11 @@ module Make(H : Hashable.S) = struct
     | Result.Ok h -> h
     | Error (_, _, _) ->
       Code_error.raise "Hashtbl.of_list_exn duplicate keys" []
+
+  let add_exn t key data =
+    match find t key with
+    | None -> set t key data
+    | Some _ -> Code_error.raise "Hastbl.add_exn: key already exists" []
```
Comment (aalekseyev) [inline src/stdune/hashtbl.ml]: Not showing the key here makes for very poor error messages, and you're removing some of the existing better error messages too.

Maybe we should keep the custom error messages for now?

File: src/alias.ml [hunk 2]
```diff
@@ -83,7 +83,7 @@ let standard_aliases = Hashtbl.create 7
 let is_standard name = Hashtbl.mem standard_aliases name
 
 let make_standard name =
-  Hashtbl.add standard_aliases name ();
+  Hashtbl.add_exn standard_aliases name ();
```
Comment (aalekseyev) [inline src/alias.ml]: `set` seems better here.

File: src/dune_project.ml [hunk 3]
```diff
@@ -344,12 +344,9 @@ module Extension = struct
 
   let register ?(experimental=false) syntax stanzas arg_to_dyn =
     let name = Syntax.name syntax in
-    if Hashtbl.mem extensions name then
```
Comment (aalekseyev) [inline src/dune_project.ml]: Let's not remove a custom error message.

File: src/versioned_file.ml [hunk 4]
```diff
@@ -42,10 +42,7 @@ module Make(Data : sig type t end) = struct
 
     let register syntax data =
       let name = Syntax.name syntax in
-      if Hashtbl.mem langs name then
-        Code_error.raise "Versioned_file.Lang.register: already registered"
```
Comment (aalekseyev) [inline src/versioned_file.ml]: Custom error message.

File: src/build_system.ml [hunk 5]
```diff
@@ -483,16 +483,7 @@ let get_dir_triage t ~dir =
         Need_step2
     end
 
-let add_spec_exn t fn rule =
```
Comment (aalekseyev) [inline src/build_system.ml]: Custom error message.

File: src/stdune/hashtbl_intf.ml [hunk 6]
```diff
@@ -1,7 +1,9 @@
 module type S = sig
   include MoreLabels.Hashtbl.S
 
-  val add : 'a t -> key -> 'a -> unit
+  val set : 'a t -> key -> 'a -> unit
+
+  val add_exn : 'a t -> key -> 'a -> unit
```
Comment (aalekseyev) [inline src/stdune/hashtbl_intf.ml]: Don't you want a version of `add` that returns a `Result`, by the way? I think that will be useful e.g. to restore custom error messages.

File: src/alias.ml (hunk 2 above)
Comment (rgrinberg) [reply src/alias.ml]: Why? it shouldn't be possible to override an alias after it's been set.

File: src/stdune/hashtbl_intf.ml (hunk 6 above)
Comment (rgrinberg) [reply src/stdune/hashtbl_intf.ml]: Indeed. I was going to add it but there's nowhere to use it yet, and I wanted to see what you thought of the interface i proposed.

**Comment by rgrinberg** [comment]
Regarding the custom error messages:

Most of them are off the form where we just print the missing key and all the available keys. I have a plan to make `Hashtbl.Make` require a `to_dyn` function so that we have decent error messages even if we don't write custom error messages. Yes, those error messages will be a little worse, but I don't think it will matter so much. I can keep the error messages for now, but I think we should consider shedding that weight if we have a "good enough" error by default.

I'm also going to replace all uses of polymorphic hash tables.

File: src/alias.ml (hunk 2 above)
Comment (aalekseyev) [reply src/alias.ml]: It's OK to do that though: the operation is idempotent. Well, either way seems fine.

**Comment by aalekseyev** [comment]
I see.

I think it's useful to have custom error messages even if add_exn gives something reasonable by default. It's a good idea to write down somewhere the reason you believe the exception shouldn't happen. Maybe it's fine to turn a custom error message into a comment, but is that really better?

And yeah, I'd prefer in any case if this happened *after* the default hashtable errors are improved.

File: src/alias.ml (hunk 2 above)
Comment (rgrinberg) [reply src/alias.ml]: It's ok, but it's still probably unintended.

**Comment by rgrinberg** [comment]
@aalekseyev I've brought back all custom error messages.

Is my proposed interface for an `add` function acceptable? I can add it here or in a follow up PR.

**Comment by aalekseyev** [comment]
I don't remember seeing an exact type, but anything returning a variant (with or without payload in the error case) is good as far as I'm concerned.

**Comment by rgrinberg** [comment]
@aalekseyev Ah, I already added it here.

**Comment by aalekseyev** [comment]
Looks great! Thanks.

**Pull request merged**